cmd/gazelle/fix-update.go

```diff
@@ -322,7 +322,13 @@ func runFixUpdate(wd string, cmd command, args []string) (err error) {
 			mappedKinds    []config.MappedKind
 			mappedKindInfo = make(map[string]rule.KindInfo)
 		)
-		for _, r := range gen {
+		// We apply map_kind to all rules, including pre-existing ones.
+		var allRules []*rule.Rule
+		allRules = append(allRules, gen...)
+		if f != nil {
+			allRules = append(allRules, f.Rules...)
+		}
```

Comment:
Question: does the append ordering matter here?

I.e. if in `foo/BUILD` I were to map `a -> b` and in `foo/bar/BUILD` I were to map `b -> c`, what would be generated in `foo/bar/tee/BUILD`?

Reply:
That's an interesting question! I think ordering of rules doesn't matter here, but there is an interesting edge-case nonetheless:

For each rule, we look up in a pre-generated config for the `map_kind` directive, which was generated in the ordinary way for config. So the config object for `foo/bar/tee/BUILD` will contain: `a -> b` and `b -> c` but _doesn't_ resolve a transitive `a -> c`. Given this is a static pre-generated config, and the mappings aren't affected by the order the rules are consulted, I don't think there's an ordering problem here.

This means for things we _generate_, if we generate an `a` we'll end up with a `b`, and if we generate a `b` we'll end up with a `c`.

For things which _already existed_, if there was an `a` we'll end up with a `b`, and if there was a `b` we'll end up with a `c`. Which means that if you have an `a`, and run gazelle twice, you'll end up with different results after each generation.

I think if we wanted to fix this, we'd probably add a loop around the `if repl, ok := c.KindMap[r.Kind()]; ok {` to keep mapping until we reach an unmapped rule, or until we detect a loop.

I've pushed https://github.com/bazelbuild/bazel-gazelle/pull/1425/commits/0c06eb2321d1691a14a1d47915ee3255dec0df73 which does this transitive mapping, and which errors if a loop was encountered. I can imagine situations where people may have loops in their `map_kind` directives today, so this _may_ be a breaking change to some.